### ISSUE TITLE
About camera pose in Gobjaverse

### ISSUE THREAD

**Issue by wenqsun** [3DTopia/LGM#30]
Thanks for your work!

I have downloaded the Gobajaverse dataset, and got .json file like below:
![image](https://github.com/3DTopia/LGM/assets/93043187/38b62cca-edb7-42c3-8119-681a4b8d69f7)

the coordinate system should be opencv, so I use the code below to transfer the matrix to the opengl system
```
camera_matrix = torch.eye(4)
camera_matrix[:3, 0] = torch.tensor(json_content["x"])
camera_matrix[:3, 1] = -torch.tensor(json_content["y"])
camera_matrix[:3, 2] = -torch.tensor(json_content["z"])
camera_matrix[:3, 3] = torch.tensor(json_content["origin"])
c2w = camera_matrix
c2w = torch.tensor(c2w, dtype=torch.float32).reshape(4, 4)
```
But I am not sure how I can change the proj_matrix, I can compute the fovy should be around 40 based on the rendering json file, but for znear and zfar, I have no information about that. Based on the experimental results, I think there may exist some problems in my dataloader part, especially the camera transformation part.

Maybe I have some misunderstandings about the `provider_objaverse.py`. I am wondering if you can provide some suggestions about this part. Thanks!

**Comment by ashawkey** [comment]
@wenqsun Hi, this is my camera loading part for g-objaverse:
```python
        for vid in vids:

            try:
                tar_path = os.path.join(self.opt.data_path, f"{uid}.tar")
                uid_last = uid.split('/')[1]

                image_path = os.path.join(uid_last, 'campos_512_v4', f"{vid:05d}/{vid:05d}.png")
                meta_path = os.path.join(uid_last, 'campos_512_v4', f"{vid:05d}/{vid:05d}.json")
                # albedo_path = os.path.join(uid_last, 'campos_512_v4', f"{vid:05d}/{vid:05d}_albedo.png") # black bg...
                # mr_path = os.path.join(uid_last, 'campos_512_v4', f"{vid:05d}/{vid:05d}_mr.png")
                nd_path = os.path.join(uid_last, 'campos_512_v4', f"{vid:05d}/{vid:05d}_nd.exr")
                
                with tarfile.open(tar_path, 'r') as tar:
                    with tar.extractfile(image_path) as f:
                        image = np.frombuffer(f.read(), np.uint8)
                    with tar.extractfile(meta_path) as f:
                        meta = json.loads(f.read().decode())
                    with tar.extractfile(nd_path) as f:
                        nd = np.frombuffer(f.read(), np.uint8)

                image = torch.from_numpy(cv2.imdecode(image, cv2.IMREAD_UNCHANGED).astype(np.float32) / 255) # [512, 512, 4] in [0, 1]

                c2w = np.eye(4)
                c2w[:3, 0] = np.array(meta['x'])
                c2w[:3, 1] = np.array(meta['y'])
                c2w[:3, 2] = np.array(meta['z'])
                c2w[:3, 3] = np.array(meta['origin'])
                c2w = torch.tensor(c2w, dtype=torch.float32).reshape(4, 4)
                
                nd = cv2.imdecode(nd, cv2.IMREAD_UNCHANGED).astype(np.float32) # [512, 512, 4] in [-1, 1]
                normal = nd[..., :3] # in [-1, 1], bg is [0, 0, 1]
                depth = nd[..., 3] # in [0, +?), bg is 0

                # rectify normal directions
                normal = normal[..., ::-1]
                normal[..., 0] *= -1
                normal = torch.from_numpy(normal.astype(np.float32)).nan_to_num_(0) # there are nans in gt normal... 
                depth = torch.from_numpy(depth.astype(np.float32)).nan_to_num_(0)
                
            except Exception as e:
                # print(f'[WARN] dataset {uid} {vid}: {e}')
                continue
            
            # blender world + opencv cam --> opengl world & cam
            c2w[1] *= -1
            c2w[[1, 2]] = c2w[[2, 1]]
            c2w[:3, 1:3] *= -1 # invert up and forward direction

            image = image.permute(2, 0, 1) # [4, 512, 512]
            mask = image[3:4] # [1, 512, 512]

            image = image[:3] * mask + (1 - mask) # [3, 512, 512], to white bg

            image = image[[2,1,0]].contiguous() # bgr to rgb

            normal = normal.permute(2, 0, 1) # [3, 512, 512]
            normal = normal * mask # to [0, 0, 0] bg

            images.append(image)
            normals.append(normal)
            depths.append(depth)
            masks.append(mask.squeeze(0))
            cam_poses.append(c2w)

            vid_cnt += 1
            if vid_cnt == self.opt.num_views:
                break

        if vid_cnt < self.opt.num_views:
            print(f'[WARN] dataset {uid}: not enough valid views, only {vid_cnt} views found!')
            n = self.opt.num_views - vid_cnt
            images = images + [images[-1]] * n
            normals = normals + [normals[-1]] * n
            depths = depths + [depths[-1]] * n
            masks = masks + [masks[-1]] * n
            cam_poses = cam_poses + [cam_poses[-1]] * n
          
        images = torch.stack(images, dim=0) # [V, 3, H, W]
        normals = torch.stack(normals, dim=0) # [V, 3, H, W]
        depths = torch.stack(depths, dim=0) # [V, H, W]
        masks = torch.stack(masks, dim=0) # [V, H, W]
        cam_poses = torch.stack(cam_poses, dim=0) # [V, 4, 4]

        # normalized camera feats as in paper (transform the first pose to a fixed position)
        radius = torch.norm(cam_poses[0, :3, 3])
        cam_poses[:, :3, 3] *= self.opt.cam_radius / radius
        transform = torch.tensor([[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, self.opt.cam_radius], [0, 0, 0, 1]], dtype=torch.float32) @ torch.inverse(cam_poses[0])
        cam_poses = transform.unsqueeze(0) @ cam_poses  # [V, 4, 4]
```

configs should be modified too:
```
# fovy (fixed for gobj)
fovy: float = 39.6
```

the normal should be rotated later too if you are using it:
```python
# rotate normal!
V, _, H, W = normal_final.shape # [1, 3, h, w]
normal_final = (transform[:3, :3].unsqueeze(0) @ normal_final.permute(0, 2, 3, 1).reshape(-1, 3, 1)).reshape(V, H, W, 3).permute(0, 3, 1, 2).contiguous()
```

**Comment by ashawkey** [comment]
zfar and znear are not that important, you could just set it to 0.01 and 1000.

**Comment by wenqsun** [comment]
Thanks for your reply!

The main part is similar to mine, and I also use the same transformation as you provide in `provider_objaverse.py`
```
cam_poses[:, :3, 1:3] *= -1 # invert up & forward direction
# cameras needed by gaussian rasterizer
cam_view = torch.inverse(cam_poses).transpose(1, 2) # [V, 4, 4]
cam_view_proj = cam_view @ self.proj_matrix # [V, 4, 4]
cam_pos = -cam_poses[:, :3, 3] # [V, 3]
```
However, during the training process, I find that the 3d gaussians and rendered images are very wired
![image](https://github.com/3DTopia/LGM/assets/93043187/752b4ff6-9420-4242-a0bc-5b6375afd237)
![image](https://github.com/3DTopia/LGM/assets/93043187/9c0f2e0c-a221-4fc5-9fe1-cb20afb4a269)
I have check the camera transformation process, it seems no problem (not very sure). Therefore, I wonder if there are some tricks or training settings ignored by me. Thanks!

**Comment by ashawkey** [comment]
This looks strange... Could you post some of the log images (including 4 input views and 4 output views) saved by `main.py`?

**Comment by wenqsun** [comment]
Of course, here are the recorded images (4 input and 4 output)
![image](https://github.com/3DTopia/LGM/assets/93043187/e101691a-69db-4398-899a-d50f9d63a9c8)

**Comment by ashawkey** [comment]
It seems pretty good to me? How do you get the bad Gaussians in the first image?

**Comment by wenqsun** [comment]
I also tried to fine-tune the pre-trained model based on the provided model weight, the result shows normal, but later the loss increase around 800 iterations,

**Comment by wenqsun** [comment]
Thanks for your help!

I tried some parameters, such as larger `gradient_accumulation_steps`, and smaller `lr`. Also, I changed some original codes, one of them is below:
```
# Here I also set the bg_color of the training dataset as white 
if self.training:
      bg_color = torch.ones(3, dtype=torch.float32, device=gaussians.device)
else:
      bg_color = torch.ones(3, dtype=torch.float32, device=gaussians.device)
```
because the lpips and mse loss are computed between gt_images and rendered images, they should keep the same bg_color.

Now the whole training process works. Thanks for your help again!

**Comment by chenguolin** [comment]
Hi @wenqsun @ashawkey,

Could you please explain why we need multiple -1 on `cam_poses[:, :3, 3]` to get `cam_pos` for the Gaussian Rasterizer?
```python
cam_pos = -cam_poses[:, :3, 3]  # instead of cam_pos = cam_poses[:, :3, 3] ?
```

Thank you very much!

**Comment by ashawkey** [comment]
@wenqsun Thanks! You are right, I used white background during training too, and I have fixed the code.

**Comment by wenqsun** [comment]
Hi @chenguolin, you can check this issue https://github.com/3DTopia/LGM/issues/15

**Comment by chenguolin** [comment]
> Hi @chenguolin, you can check this issue #15

I got it. Thank you very much for the quick reply!

**Comment by l772308126** [comment]
> Now the whole training process works. Thanks for your help again!

Hi, Im using the Gobjaverse dataset too. Could you provide me more details (such as epoch, pct_start in scheduler...) and your code of the ObjaverseDataset in 'provider_objaverse.py' ? Currently, I face the problem of getting strange outputs. The first is gt and the second is pred.
gt :
![train_gt_images_71_0](https://github.com/user-attachments/assets/49557608-495b-4221-96f0-91c9ae0ed426)
pred:
![train_pred_images_71_0](https://github.com/user-attachments/assets/a373c4e1-b225-4d0f-92a0-d5a600bcd590)
some places are just blank space.

**Comment by geonyeong-park** [comment]
Hi @ashawkey @wenqsun , thank you very much for the great work, open-sourcing and valuable discussions!
I'm following the @ashawkey's g-objaverse loading code, and from my side observed that the batch sampling is very slow.

I've filtered out abnormal tar files, so there are almost no exceptions during sampling.
But, still a single batch (size 8, view 12) sampling requires about 5~8 seconds with a single GPU, which bottlenecks the training (GPU is not a bottleneck in this case).
Things become even worse when i try multi-gpu training. 

It would be a great pleasure if you can share your experience. Have you ever experienced same issues?
